File: code-examples/Python/target_tracking/kalman_filter.py
#!/usr/bin/env python

#
# Author: Fangtong Liu
# Date: 03/30/2020
#

import numpy as np


# Kalman filter class for state estimation of a linear system
class kalman_filter:

    def __init__(self, system, init):
        # KALMAN_FILTER Construct an instance of this class
        #
        # Inputs:
        #   system: linear system and noise models
        #   init:   initial state mean and covariance

        self.A = system.A  # system matrix
        self.B = system.B  # input matrix
        self.H = system.H  # measurement model
        self.Q = system.Q  # input noise covariance
        self.R = system.R  # measurement noise covariance
        self.x = init.x  # state vector
        self.Sigma = init.Sigma  # state covariance

    def prediction(self):
        # Kalman filter propagation (prediction) step

        self.x_pred = np.dot(self.A, self.x)  # + np.dot(self.B, u) # predicted state
        self.Sigma_pred = np.dot(np.dot(self.A, self.Sigma), np.transpose(self.A)) + self.Q  # predicted state covariance
        self.z_hat = np.dot(self.H, self.x_pred)  # predicted measurement
        # print("x_pred", self.x_pred)

    def correction(self, z):
        # Kalman filter correction step
        #
        # Inputs:
        #   z:  measurement

        # compute innovation statistics
        self.v = z - self.z_hat  # innovation
        self.S = np.dot(np.dot(self.H, self.Sigma_pred), np.transpose(self.H)) + self.R  # innovation covariance

        # filter gain
        self.K = np.dot(np.dot(self.Sigma_pred, np.transpose(self.H)), np.linalg.inv(self.S))  # Kalman (filter) gain

        # correct the predicted state statistics
        self.x = self.x_pred + np.dot(self.K, self.v)
        I = np.eye(np.shape(self.x)[0])
        temp = np.transpose(I - np.dot(self.K, self.H))
        self.Sigma = np.dot(np.dot(temp, self.Sigma_pred), temp.T) + np.dot(np.dot(self.K, self.R), np.transpose(self.K))  # Joseph update form

File: code-examples/MATLAB/target_tracking/@kalman_filter/kalman_filter.m
classdef kalman_filter < handle
    % Kalman filter class for state estimation of a linear system
    %
    %   Author: Maani Ghaffari Jadidi
    %   Date:   11/24/2018
    
    properties
        A;              % system matrix
        B;              % input matrix
        H;              % measurement model
        S;              % innovation covariance
        v;              % innovation
        x;              % state vector
        z_hat;          % predicted measurement
        Sigma;          % state covariance
        x_pred;         % predicted state
        Sigma_pred;     % predicted state covariance
        Q;              % input noise covariance
        R;              % measurement noise covariance
        K;              % Kalman (filter) gain
        
    end
    
    methods
        function obj = kalman_filter(system, init)
            % KALMAN_FILTER Construct an instance of this class
            %
            %   Inputs:
            %       system          - linear system and noise models
            %       init            - initial state mean and covariance
            
            obj.A = system.A;
            obj.B = system.B;
            obj.H = system.H;
            obj.Q = system.Q;
            obj.R = system.R;
            obj.x = init.x;
            obj.Sigma = init.Sigma;
        end
        
        function prediction(obj)
            % Kalman filter propagation (prediction) step
            obj.x_pred = obj.A * obj.x; % + obj.B * u;
            obj.Sigma_pred = obj.A * obj.Sigma * obj.A' + obj.Q;
            obj.z_hat = obj.H * obj.x_pred;
        end
        
        function correction(obj, z)
            % Kalman filter correction step
            %
            %   Inputs:
            %       z          - measurement
            
            % compute innovation statistics
            obj.v = z - obj.z_hat;
            obj.S = obj.H * obj.Sigma_pred * obj.H' + obj.R;
            
            % filter gain
            obj.K = obj.Sigma_pred * obj.H' * (obj.S \ eye(size(obj.S)));
            
            % correct the predicted state statistics
            obj.x = obj.x_pred + obj.K * obj.v;
            I = eye(size(obj.x));
            obj.Sigma = ...
                (I - obj.K * obj.H) * obj.Sigma_pred * (I - obj.K * obj.H)' ...
                    + obj.K * obj.R * obj.K'; % Joseph update form
        end
    end
end